src/nvme/linux.c

```diff
@@ -656,11 +794,69 @@ long nvme_lookup_keyring(const char *keyring)
 	}
 	return keyring_id;
 }
+
+long nvme_insert_tls_key(const char *keyring, const char *key_type,
+			 const char *hostnqn, const char *subsysnqn, int hmac,
+			 unsigned char *configured_key, int key_len)
+{
+	key_serial_t keyring_id, key = 0;
+	char *identity;
+	unsigned char *psk;
+
+	keyring_id = nvme_lookup_keyring(keyring);
+	if (keyring_id <= 0)
+		return 0;
+
+	identity = malloc(strlen(hostnqn) + strlen(subsysnqn) + 12);
+	if (!identity)
+		return 0;
+
+	sprintf(identity, "NVMeR%02d %s %s", hmac, hostnqn, subsysnqn);
+
+	psk = derive_nvme_keys(hostnqn, identity, hmac,
+			       configured_key, key_len);
+	if (!psk)
+		goto out_free_identity;
+
+	key = keyctl_search(keyring_id, key_type, identity, 0);
+	if (key >= 0) {
+		long err = keyctl_update(key, psk, key_len);
+		if (err) {
+			nvme_msg(NULL, LOG_ERR,
+				 "updating %s key '%s' failed\n",
+				 key_type, identity);
+			key = 0;
+		}
+	} else {
+		key = add_key(key_type, identity,
+			      psk, key_len, keyring_id);
+		if (key < 0) {
+			nvme_msg(NULL, LOG_ERR,
+				 "adding %s key '%s' failed, error %d\n",
+				 key_type, identity, errno);
+			key = 0;
+		}
+	}
+	free(psk);
+out_free_identity:
+	free(identity);
+	return key;
+}
+
 #else
 long nvme_lookup_keyring(const char *keyring)
 {
 	nvme_msg(NULL, LOG_ERR, "key operations not supported; "\
 		 "recompile with keyutils support.\n");
 	return 0;
 }
+
+long nvme_insert_tls_key(const char *keyring, const char *key_type,
+			 const char *hostnqn, const char *subsysnqn, int hmac,
+			 unsigned char *configured_key, int key_len)
+{
+	if (derive_nvme_keys(NULL, NULL, 0, NULL, 0))
+		return 0;
+	return 0;
```

Comment:
Twice `return 0`

Reply:
Nasty dummy functions. Probably time for the __unused__ attribute.

Reply:
Removed with the update.